Misc. fixes to datatypes modules

jdavcs opened the pull request:
This is extracted from #14007. I'm splitting that PR into 2 parts: one is for bug fixes and refactoring that came up in the process of adding type hints (this one), the other one is type hints only. There are a lot of type hints, and I think splitting this in two PRs  makes sense. Reviewing should be much easier too.



## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jdavcs committed:
Do not instantiate MarkdownFormatHelpers

No need for a separate instance or even a class: it's just functions;
also instead of passing references, just import as needed.

jdavcs committed:
Drop unnecessary check: data is not a string

jdavcs committed:
Check for None: TarFile.extractfile can return None

https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extractfile

jdavcs committed:
Fix bug: don't drop args in call to parent

jdavcs committed:
Remove commented out classes used by outdated versions of PlantTribes tools

Commented classes supported versions 1.0.0 - 1.0.2.
Current version is 1.0.4 from 2019.

jdavcs committed:
Fix bug in get_file_peek + add type hints

On error, slicing was applied to string (default message), not list of lines read
from the file.

jdavcs committed:
Fix bug: no `valid` method defined on DatatypeValidation

Must have been a typo: should be `validated`.

jdavcs committed:
Add property stubs for required attributes

Bug uncovered by mypy. Attrs must be defined in subclasses (as per
docstring comment).

jdavcs committed:
Fix mypy error in isa.py

jdavcs committed:
Correct file permissions of goldenpath.py (remove exec)

jdavcs committed:
Fix bug in anvio.py: check if dataset before accessing its attribute

mvdbeek commented on lib/galaxy/datatypes/anvio.py:
```diff
@@ -36,17 +36,18 @@ def generate_primary_file(self, dataset=None):
                     opt_text = " (optional)"
                 missing_text = ""
-                if not os.path.exists(os.path.join(dataset.extra_files_path, composite_name)):
+                if dataset and not os.path.exists(os.path.join(dataset.extra_files_path, composite_name)):
```
Can dataset ever be None ? Maybe remove the default value instead ?

jdavcs replied:
The method is inherited from the base class: AnvioComposite < Html < Text < Data. We can't remove the default in Data since it is indeed optional in many of its subclasses; but if we remove it only here, mypy will complain after type hints are added (can't break the signature of a parent class method). 
Another option is to add an `assert dataset` statement, or a mypy ignore comment; I think explicitly checking the value is the easiest and safest option here, but we could use either one instead.

mvdbeek replied:
Yeah, but all the calling paths provide dataset afaict, what's the point then to have it optional and check for cases that can never happen?

jdavcs replied:
Yes, I think you're right - it doesn't quite make sense. I, actually, came across multiple similar cases in `datatypes` while adding typing. There's some method `do_foo` that is defined in a parent class and takes a whole bunch of arguments; and then there are lots of subclasses which define the same method but, in most cases, don't need most of the arguments defined in the parent. So, I think, the contract is wrong: instead of "subclass of parent must define method `do_foo` with these specific arguments" it should be "subclass of parent must/may define method `do_foo`". Thus, pulling out all such methods into an abstract base class, giving them a flexible (`*args, **kwgs`) signature, and then 
 overriding as needed with specific sets of parameters should make it both correct and clean. I'll do one follow-up PR for all such cases.

mvdbeek commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -3357,7 +3357,8 @@ def set_meta(self, dataset, overwrite=True, **kwd):
                 with tarfile.open(dataset.file_name, "r") as temptar:
                     pg_version_file = temptar.extractfile("postgresql/db/PG_VERSION")
-                    dataset.metadata.version = util.unicodify(pg_version_file.read()).strip()
+                    if pg_version_file:
```
maybe raise an exception if there's no pg_version_file ?

jdavcs replied:
Done.

mvdbeek commented on lib/galaxy/datatypes/genetics.py:
```diff
@@ -787,17 +787,16 @@ def get_peek(self, dataset):
         try:
             with open(filename) as f:
-                h = f.readlines()
+                lines = f.readlines()
```
Seems like the idea is to concatenate the first 5 lines, maybe we should do that without reading in the whole file.

jdavcs replied:
Done.

mvdbeek reviewed (approved):
Nice!

jdavcs committed:
Raise exception in set_meta if PG_VERSION file not found

jdavcs committed:
Read 5 lines instead of file

jdavcs committed:
Make dataset parameter required

It's never None for this class, so having the parameter as optional is
semantically incorrect. This, however, does violate the Liskov
substitution principle (subclass method signature more strict than
parent definition), but we can fix this (as well as many other such
cases in datatypes) by slightly modifying the object hierarchy. Should
be submitted in a follow-up PR.

jdavcs committed:
Fix f-string error

jdavcs committed:
Fix linting
